subsys/bluetooth/mesh/app_keys.c

```diff
@@ -0,0 +1,506 @@
+{
+	for (int i = 0; i < ARRAY_SIZE(apps); i++) {
+		if (apps[i].app_idx == app_idx) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+ssize_t bt_mesh_apps_get(uint16_t net_idx, uint16_t app_idxs[], off_t skip,
+			 size_t max)
+{
+	size_t count = 0;
+
+	for (int i = 0; i < ARRAY_SIZE(apps); i++) {
+		struct bt_mesh_app *app = &apps[i];
+
+		if (app->app_idx == BT_MESH_KEY_UNUSED) {
+			continue;
+		}
+
+		if (net_idx != BT_MESH_KEY_ANY && app->net_idx != net_idx) {
+			continue;
+		}
+
+		if (skip) {
+			skip--;
+			continue;
+		}
+
+		if (count >= max) {
+			return -ENOMEM;
+		}
+
+		app_idxs[count++] = app->app_idx;
+	}
+
+	return count;
+}
+
+void bt_mesh_app_key_cb_register(struct bt_mesh_app_key_cb *cb)
+{
+	sys_slist_append(&app_key_cbs, &cb->n);
+}
+
+int bt_mesh_keys_resolve(struct bt_mesh_msg_ctx *ctx,
+			 struct bt_mesh_subnet **sub,
+			 const uint8_t *app_key[16], uint8_t *aid)
+{
+	struct bt_mesh_app *app = NULL;
+
+	if (BT_MESH_IS_DEV_KEY(ctx->app_idx)) {
+		/* With device keys, the application has to decide which subnet
+		 * to send on.
+		 */
+		*sub = bt_mesh_subnet_get(ctx->net_idx);
+		if (!*sub) {
+			BT_WARN("Unknown NetKey 0x%03x", ctx->net_idx);
+			return -EINVAL;
+		}
+
+		if (ctx->app_idx == BT_MESH_KEY_DEV_REMOTE &&
+		    !bt_mesh_elem_find(ctx->addr)) {
+			struct bt_mesh_cdb_node *node;
+
+			if (!IS_ENABLED(CONFIG_BT_MESH_CDB)) {
+				BT_WARN("No DevKey for 0x%04x", ctx->addr);
+				return -EINVAL;
+			}
+
+			node = bt_mesh_cdb_node_get(ctx->addr);
+			if (!node) {
+				BT_WARN("No DevKey for 0x%04x", ctx->addr);
+				return -EINVAL;
+			}
+
+			*app_key = node->dev_key;
+		} else {
+			*app_key = bt_mesh.dev_key;
+		}
+
+		*aid = 0;
+		return 0;
+	}
+
+	app = app_get(ctx->app_idx);
+	if (!app) {
+		BT_WARN("Unknown AppKey 0x%03x", ctx->app_idx);
+		return -EINVAL;
+	}
+
+	*sub = bt_mesh_subnet_get(app->net_idx);
+	if (!*sub) {
+		BT_WARN("Unknown NetKey 0x%03x", app->net_idx);
+		return -EINVAL;
+	}
+
+	if ((*sub)->kr_phase == BT_MESH_KR_PHASE_2 && app->updated) {
+		*aid = app->keys[1].id;
+		*app_key = app->keys[1].val;
+	} else {
+		*aid = app->keys[0].id;
+		*app_key = app->keys[0].val;
+	}
+
+	return 0;
+}
+
+uint16_t bt_mesh_app_key_find(bool dev_key, uint8_t aid,
+			      struct bt_mesh_net_rx *rx,
+			      int (*cb)(struct bt_mesh_net_rx *rx,
+					const uint8_t key[16], void *cb_data),
+			      void *cb_data)
+{
+	int err, i;
+
+	if (dev_key) {
+		/* Attempt remote dev key first, as that is only available for
+		 * provisioner devices, which normally don't interact with nodes
+		 * that know their local dev key.
+		 */
+		if (IS_ENABLED(CONFIG_BT_MESH_CDB) &&
+		    rx->net_if != BT_MESH_NET_IF_LOCAL) {
+			struct bt_mesh_cdb_node *node;
+
+			node = bt_mesh_cdb_node_get(rx->ctx.addr);
+			if (node && !cb(rx, node->dev_key, cb_data)) {
+				return BT_MESH_KEY_DEV_REMOTE;
+			}
+		}
+
+		err = cb(rx, bt_mesh.dev_key, cb_data);
+		if (!err) {
+			return BT_MESH_KEY_DEV_LOCAL;
+		}
+
```

Comment:
should add check (devkey with unicast address) here?

Reply:
Added